metricbeat/mb/testing/modules.go

```diff
@@ -262,17 +274,16 @@ func RunPushMetricSetV2(timeout time.Duration, waitEvents int, metricSet mb.Push
 	}()
 
 	timeoutC := time.After(timeout)
-	for {
+	for done := false; !done; {
 		select {
 		case <-termC:
-			return r.Events
+			done = true
 		case <-timeoutC:
-			return r.Events
+			done = true
 		default:
-			if waitEvents != 0 && waitEvents <= r.eventCount() {
-				return r.Events
-			}
 			time.Sleep(100 * time.Millisecond)
+			done = waitEvents != 0 && waitEvents <= r.eventCount()
 		}
 	}
+	return r.Events()
 }
```

Comment:
May I suggest an alternative approach that uses channels instead of locks. How does this look:

```
// capturingReporterV2 stores all the events and errors from a metricset's
// Run method.
type capturingReporterV2 struct {
	done    chan struct{}
	eventsC chan mb.Event
}

// Event stores the passed-in event into the events array
func (r *capturingReporterV2) Event(event mb.Event) bool {
	r.eventsC <- event
	return true
}

// Error stores the given error into the errors array.
func (r *capturingReporterV2) Error(err error) bool {
	r.eventsC <- mb.Event{Error: err}
	return true
}

// Done returns the Done channel for this reporter.
func (r *capturingReporterV2) Done() <-chan struct{} {
	return r.done
}

// RunPushMetricSetV2 run the given push metricset for the specific amount of
// time and returns all of the events and errors that occur during that period.
func RunPushMetricSetV2(timeout time.Duration, waitEvents int, metricSet mb.PushMetricSetV2) []mb.Event {
	var (
		r      = &capturingReporterV2{done: make(chan struct{}), eventsC: make(chan mb.Event, 1)}
		wg     sync.WaitGroup
		events []mb.Event
	)

	// Producer
	wg.Add(1)
	go func() {
		defer wg.Done()
		defer close(r.eventsC)
		metricSet.Run(r)
	}()

	// Consumer
	wg.Add(1)
	go func() {
		defer wg.Done()
		defer close(r.done)

		timer := time.NewTimer(timeout)
		defer timer.Stop()
		for {
			select {
			case <-timer.C:
				return
			case e := <-r.eventsC:
				events = append(events, e)
				if len(events) >= waitEvents {
					return
				}
			}
		}
	}()

	wg.Wait()
	return events
}
```

Reply:
One thing I liked about the original code is that it allows the metricSet to generate more events than expected and this can be detected by tests. But only one test was relying on this

Reply:
@adriansr A bit late to the game, sorry. But can you elaborate on this how we changed the behaviour?

Reply:
I think what he was referring to is that in my version I forgot to allow for `waitEvents == 0` which indicates that the only bound on the running time is the timeout (a user doesn't have to specify the exact number of events to receive).

In the final version @adriansr changed the `if len(events) >= waitEvents {` to `if waitEvents > 0 && len(events) >= waitEvents {`.